packages/core/src/__tests__/utils.test.ts

```diff
@@ -0,0 +1,172 @@
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import { mergeQueries } from '..'
+
+describe('mergeQueries', () => {
+  it('merges query with empty query', () => {
+    const q1 = { name: 'john', age: { $gt: 42 } }
+    const q2 = {}
+    const res = { name: 'john', age: { $gt: 42 } }
+
+    expect(mergeQueries(q1, q2)).toEqual(res)
+    expect(mergeQueries(q2, q1)).toEqual(res)
+  })
+
+  it('merges query with different fields', () => {
+    const q1 = { name: 'john' }
+    const q2 = { age: { $gt: 42 } }
+    const res = { name: 'john', age: { $gt: 42 } }
+
+    expect(mergeQueries(q1, q2)).toEqual(res)
+    expect(mergeQueries(q2, q1)).toEqual(res)
+  })
+
+  it('merges equal field values', () => {
+    expect(mergeQueries({ value: 42 }, { value: 42 })).toEqual({ value: 42 })
+    expect(mergeQueries({ value: 42 }, { value: 42 })).toEqual({ value: 42 })
+  })
+
+  it('does not merge different field values', () => {
+    const q1 = { value: 42 }
+    const q2 = { value: 'true' }
+    const res = { value: { $in: [] } }
+    expect(mergeQueries(q1, q2)).toEqual(res)
+    expect(mergeQueries(q2, q1)).toEqual(res)
+  })
+
+  it('merges predicate with predicate', () => {
+    expect(mergeQueries({ age: { $in: [41, 42] } }, { age: { $in: [42, 43] } })).toEqual({ age: { $in: [42] } })
+    expect(mergeQueries({ age: { $in: [42, 43] } }, { age: { $in: [41, 42] } })).toEqual({ age: { $in: [42] } })
+
+    expect(mergeQueries({ age: { $nin: [42] } }, { age: { $nin: [42] } })).toEqual({ age: { $nin: [42] } })
+    expect(mergeQueries({ age: { $nin: [42] } }, { age: { $nin: [42] } })).toEqual({ age: { $nin: [42] } })
+
+    expect(mergeQueries({ age: { $lt: 45 } }, { age: { $lt: 42 } })).toEqual({ age: { $lt: 42 } })
+    expect(mergeQueries({ age: { $lt: 42 } }, { age: { $lt: 45 } })).toEqual({ age: { $lt: 42 } })
+
+    expect(mergeQueries({ age: { $gt: 40 } }, { age: { $gt: 42 } })).toEqual({ age: { $gt: 42 } })
+    expect(mergeQueries({ age: { $gt: 42 } }, { age: { $gt: 40 } })).toEqual({ age: { $gt: 42 } })
+
+    expect(mergeQueries({ age: { $lte: 45 } }, { age: { $lte: 42 } })).toEqual({ age: { $lte: 42 } })
+    expect(mergeQueries({ age: { $lte: 42 } }, { age: { $lte: 45 } })).toEqual({ age: { $lte: 42 } })
+
+    expect(mergeQueries({ age: { $gte: 40 } }, { age: { $gte: 42 } })).toEqual({ age: { $gte: 42 } })
+    expect(mergeQueries({ age: { $gte: 42 } }, { age: { $gte: 40 } })).toEqual({ age: { $gte: 42 } })
+
+    expect(mergeQueries({ age: { $ne: 42 } }, { age: { $ne: 42 } })).toEqual({ age: { $nin: [42] } })
```

Comment:
We need to be aware performance are same in this case.

Reply:
I modified this to preserve original operator when possible.